File: s-des/Class/Mapper.cs
using s_des.Constants;

namespace s_des.Class;

// map permutations to an enum of permutations

public static class Mapper
{
    public static BitBuffer MapPermutation(IReadOnlyList<int> permutation, BitBuffer buffer)
    {
        var result = new int[permutation.Count];
        for (var i = 0; i < permutation.Count; i++) result[i] = buffer.Buffer[permutation[i] - 1];

        return new BitBuffer(result);
    }

    public static BitBuffer MapWithSboxs(BitBuffer buffer)
    {
        // mapping left
        int[] mapping = {0, 3, 1, 2};
        var left = MapWithSbox(buffer.Left, SBox.S0, mapping);

        var right = MapWithSbox(buffer.Right, SBox.S1, mapping);

        return new BitBuffer(left.Concat(right).ToArray());
    }

    private static IEnumerable<int> MapWithSbox(BitBuffer buffer, int[,] sBox, int[] mapping)
    {
        // convert to 2d array
        var row = Convert.ToInt32($"{buffer.Buffer[mapping[0]]}{buffer.Buffer[mapping[1]]}", 2);
        var column = Convert.ToInt32($"{buffer.Buffer[mapping[2]]}{buffer.Buffer[mapping[3]]}", 2);
        // convert to binary form sBox[row, column]
        var s = Convert.ToString(sBox[row, column], 2);
        if (s.Length == 1) s = "0" + s;
        // convert to int array
        var result = new int[s.Length];
        for (var i = 0; i < s.Length; i++) result[i] = Convert.ToInt32(s[i].ToString());
        return result;
    }
}

File: s-des typescript/src/class/Mapper.ts
import BitBuffer from "./BitBuffer";
import { SBox } from "../constants/sbox";

export class Mapper {
  public static MapPermutation(
    permutation: ReadonlyArray<number>,
    buffer: BitBuffer
  ): BitBuffer {
    const result = new Array<number>(permutation.length);
    for (let i = 0; i < permutation.length; i++) {
      result[i] = buffer.buffer[permutation[i] - 1];
    }

    return new BitBuffer(result);
  }

  public static MapWithSboxs(buffer: BitBuffer): BitBuffer {
    const mapping = [0, 3, 1, 2];
    const left = Mapper.MapWithSbox(buffer.left!, SBox.S0, mapping);
    const right = Mapper.MapWithSbox(buffer.right!, SBox.S1, mapping);
    const concatenated = left.concat(right);
    return new BitBuffer(concatenated);
  }

  private static MapWithSbox(
    buffer: BitBuffer,
    sBox: number[][],
    mapping: number[]
  ): number[] {
    const row = parseInt(
      `${buffer.buffer[mapping[0]]}${buffer.buffer[mapping[1]]}`,
      2
    );
    const column = parseInt(
      `${buffer.buffer[mapping[2]]}${buffer.buffer[mapping[3]]}`,
      2
    );
    let s = sBox[row][column].toString(2);
    if (s.length === 1) s = `0${s}`;
    const result = new Array<number>(s.length);
    for (let i = 0; i < s.length; i++) {
      result[i] = parseInt(s[i]);
    }

    return result;
  }
}
